### ISSUE TITLE
Is there any way to use the HttpClientFactory for providing the HttpClients?

### ISSUE THREAD

**Issue by emorell96** [RogueException/Discord.Net#1805]
So I have being trying to understand how the library, in particular the .net.rest part of the library handles creation of the HttpClient objects, and it seems I might have found the answer in [Line 31 of DefaultRestClient](https://github.com/discord-net/Discord.Net/blob/1b64d19c845cb7c612a1c52288c8b44cff605105/src/Discord.Net.Rest/Net/DefaultRestClient.cs#L31).

If I understand the code (which I have browsed quite a bit, but I still wanted to confirm with you all since you know this library 10x better), the DefaultRestClientProvider everytime it gets called, it will return a new HttpClient with a new HttpHandler.

So if I were to register a DiscordRestClient as a scoped, or transient service, every time a new instance of DiscordRestClient would be created a new HttpClient and HttpHandler would be created as well leading to exactly the issue talked about in this:

https://www.aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests

Would there be a way to use the IHttpClientFactory to create HttpClients in this library?

Maybe a different DefaultRestClient that could use the IHttpClientFactory?

I'd be happy to contribute if this is in fact an issue.

Thanks!

**Comment by SubZero0** [comment]
> So if I were to register a DiscordRestClient as a scoped, or transient service, every time a new instance of DiscordRestClient would be created a new HttpClient and HttpHandler would be created as well leading to exactly the issue talked about in this:

Exactly, so you shouldn't. It needs to be added as a singleton as it was designed, then it will only use a single HttpClient as shown as the "fix" in the exact page (wouldn't solve the dns issues but well).

> Maybe a different DefaultRestClient that could use the IHttpClientFactory?

You can already do that by having a custom provider to achieve the same, except it still shouldn't be added as anything but  a singleton.

**Now, why singleton?** All the clients have an internal ratelimiter to prevent you from hitting them and being possibly banned by Discord for it. By creating multiple clients, you are effectively making it useless.

Basically, add it as singleton, you have this option.

**Comment by quinchs** [comment]
Closed in #1806
